src/main/java/picard/analysis/CounterManager.java

```diff
@@ -0,0 +1,189 @@
+     *
+     * @param locusPosition position in the reference sequence,
+     *                      that will be represented by index 0 of <code>Counter</code> arrays.
+     */
+    public void checkOutOfBounds(int locusPosition) {
+        if (locusPosition - offset + readLength >= arrayLength) {
+            if (locusPosition - offset < arrayLength) {
+                rebase(locusPosition);
+            } else {
+                clear();
+                offset = locusPosition;
+            }
+        }
+    }
+
+    /**
+     * Rebases inner Counter arrays so that 0 index of array represents the new locus position
+     *
+     * @param locusPosition position in the reference sequence,
+     *                      that will be represented by index 0 of <code>Counter</code> arrays.
+     */
+    private void rebase(int locusPosition) {
+        for (Counter counter : arrays) {
+            int[] array = counter.array;
+            System.arraycopy(array, locusPosition - offset, array, 0, arrayLength - locusPosition + offset);
+            Arrays.fill(array, arrayLength - locusPosition + offset, arrayLength, 0);
+        }
+        offset = locusPosition;
+    }
+
+    /**
+     * Clears all inner Counter arrays
+     */
+    public void clear() {
+        for (Counter counter : arrays) {
+            int[] array = counter.array;
+            Arrays.fill(array, 0);
+        }
+        offset = 0;
+    }
+
+    /**
+     * Creates a new Counter object and adds it to the list of managed Counters.
+     *
+     * @return <code>Counter</code>, that will be managed by current <code>CounterManager</code>
+     */
+    public Counter newCounter() {
+        Counter counter = new Counter(arrayLength);
+        arrays.add(counter);
+        return counter;
+    }
+
+    /**
+     * Class represents an integer array with methods to increment and get the values form it with respect
+     * to offset of outer <code>CounterManager</code>.
+     */
+    public class Counter {
+
+        /**
+         * Inner array to store accumulated values
+         */
+        private int[] array;
+
+        /**
+         * @param arrayLength length of the inner array
+         */
+        private Counter(int arrayLength) {
+            array = new int[arrayLength];
+        }
+
+        /**
+         * Increments value corresponding to reference sequence index
+         *
+         * @param index in the reference sequence
+         */
+        public void inc(int index) {
+            checkIndex(index);
+            array[index - offset]++;
+        }
+
+        /**
+         * @param index in the reference sequence
+         * @return value from the inner array, corresponding to input reference index
+         */
+        public int get(int index) {
+            checkIndex(index);
+            return array[index - offset];
+        }
+
+        private void checkIndex(int index) {
+            if ((index - offset) < 0 || (index - offset) >= array.length) {
```

Comment:
Yes, array access itself will throw identical exception but with not meaningful index. The real index used in array is [index – offset]. So, for getting more clear exception message we`ve perform this test here.